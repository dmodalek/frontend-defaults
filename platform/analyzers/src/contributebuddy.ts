import { directoryExists, getDependencyInstallation, DependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ContributebuddyAnalyzerResult = {
    contributebuddy: boolean;
    contributebuddyConfigFolder?: string,
    contributebuddyInstallation?: DependencyInstallation
};

export const contributebuddyAnalyzer = async (cwd: string): Promise<ContributebuddyAnalyzerResult> => {
    const CONTRIBUTE_BUDDY_FOLDER = join(cwd, '.contributebuddy');
    const contributebuddyConfigFolder = await directoryExists(CONTRIBUTE_BUDDY_FOLDER);
    const contributebuddyInstallation = await getDependencyInstallation(cwd, 'contribute-buddy');

    if (contributebuddyConfigFolder || contributebuddyInstallation.declared) {
        return {
            contributebuddy: true,
            contributebuddyConfigFolder: CONTRIBUTE_BUDDY_FOLDER,
            contributebuddyInstallation

        }
    }

    return {
        contributebuddy: false,
    };
}