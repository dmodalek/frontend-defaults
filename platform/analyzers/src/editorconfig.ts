import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type EditorconfigAnalyzerResult = {
	editorConfig: boolean;
};

export class EditorconfigAnalyzer extends Analyzer<EditorconfigAnalyzerResult> {
	async analyze(): Promise<EditorconfigAnalyzerResult> {
		const eidtorConfigExists = await fileExists(join(this.context, '.editorconfig'));

		return {
			editorConfig: eidtorConfigExists,
		};
	}
}
