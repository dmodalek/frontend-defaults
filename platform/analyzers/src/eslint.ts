import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: DependencyInstallation
};

export const eslintAnalyzer = async (cwd: string): Promise<ESLintAnalyzerResult> => {
	const doesESLintExist = await fileExists(join(cwd, '.eslintrc.js'));

	if (doesESLintExist) {
		const doesESLintIgnoreExist = await fileExists(join(cwd, '.eslintignore'));
		const eslintInstallation = await getDependencyInstallation(cwd, 'eslint');

		return {
			eslint: doesESLintExist,
			eslintIgnore: doesESLintIgnoreExist,
			eslintInstallation
		};
	}

	return {
		eslint: doesESLintExist,
	};
}