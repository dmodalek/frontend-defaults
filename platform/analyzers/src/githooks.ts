import {
	DependencyInstallation,
	fileExists,
	getDependencyInstallation,
	getPackageJSON,
	IPackage,
} from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitHooksLocation = 'rc' | 'package';

export type GitHooksAnalyzerResult = {
	gitHooks: boolean;
	gitHooksInstallation?: DependencyInstallation;
	gitHooksLocation?: GitHooksLocation;
};

type PackageWithHuskyConfig = IPackage & {
	husky: any;
};

export const gitHooksAnalyzer = async (cwd: string): Promise<GitHooksAnalyzerResult> => {
	const doesGitHooksConfigExist = await fileExists(join(cwd, '.huskyrc'));
	const gitHooksInstallation = await getDependencyInstallation(cwd, 'husky');
	const pkg: PackageWithHuskyConfig = (await getPackageJSON(cwd)) as any;

	if (doesGitHooksConfigExist || pkg.husky) {
		return {
			gitHooks: doesGitHooksConfigExist,
			gitHooksLocation: pkg.husky ? 'package' : 'rc',
			gitHooksInstallation,
		};
	}

	return {
		gitHooks: false,
	};
};
