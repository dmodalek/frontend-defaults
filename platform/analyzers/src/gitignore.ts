import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export const gitIgnoreAnalyzer = async (cwd: string): Promise<GitIgnoreAnalyzerResult> => {
	const doesGitignoreExist = await fileExists(join(cwd, '.gitignore'));

	return {
		gitignore: doesGitignoreExist,
	};
}