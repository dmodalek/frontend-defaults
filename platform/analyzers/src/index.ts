// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
