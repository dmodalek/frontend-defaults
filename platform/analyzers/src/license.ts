import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type LicenseAnalyzerResult = {
	license: boolean;
};

export class LicenseAnalyzer extends Analyzer<LicenseAnalyzerResult> {
	async analyze(): Promise<LicenseAnalyzerResult> {
		const licenseExists = await fileExists(join(this.context, 'LICENSE'));

		return {
			license: licenseExists,
		};
	}
}
