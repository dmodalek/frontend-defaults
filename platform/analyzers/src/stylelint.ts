import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type StylelintAnalyzerResult = {
	stylelint: boolean;
	stylelintIgnore?: boolean;
	stylelintInstallation?: DependencyInstallation;
};

export const stylelintAnalyzer = async (cwd: string): Promise<StylelintAnalyzerResult> => {
	const doesStyleLintExist = await fileExists(join(cwd, 'stylelint.config.js'));

	if (doesStyleLintExist) {
		const doesStyleLintIgnoreExist = await fileExists(join(cwd, '.stylelintignore'));
		const stylelintInstallation = await getDependencyInstallation(cwd, 'stylelint');

		return {
			stylelint: doesStyleLintExist,
			stylelintIgnore: doesStyleLintIgnoreExist,
			stylelintInstallation,
		};
	}

	return {
		stylelint: doesStyleLintExist,
	};
};
