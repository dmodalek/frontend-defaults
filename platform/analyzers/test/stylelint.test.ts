import { getFixtureContext } from './utils';
import { stylelintAnalyzer } from '../src/stylelint';

const FIXTURE_INSTALLED = getFixtureContext('stylelint-project');
const FIXTURE_NOT_INSTALLED = getFixtureContext('stylelint-project-not-installed');

describe('Analyzers', () => {
    describe('StylelintAnalyzer', () => {
        it('should not crash', async () => {
            expect(async () => await stylelintAnalyzer(FIXTURE_INSTALLED)).not.toThrow();
            expect(async () => await stylelintAnalyzer(FIXTURE_NOT_INSTALLED)).not.toThrow();
        });

        it('should analyze a project with installation correctly', async () => {
            const analytics = await stylelintAnalyzer(FIXTURE_INSTALLED);

            expect(analytics.stylelint).toEqual(true);
            expect(analytics.stylelintIgnore).toEqual(true);
            expect(analytics.stylelintInstallation!.declared).toEqual('9.0.0');
            expect(analytics.stylelintInstallation!.installed).toEqual('9.10.1');
            expect(analytics).toMatchSnapshot({
                stylelintInstallation: {
                    latest: expect.any(String)
                }
            });
        });

        it('should analyze a project without ignore & installation correctly', async () => {
            const analytics = await stylelintAnalyzer(FIXTURE_NOT_INSTALLED);

            expect(analytics.stylelint).toEqual(true);
            expect(analytics.stylelintIgnore).toEqual(false);
            expect(analytics.stylelintInstallation!.installed).toBeFalsy();
            expect(analytics).toMatchSnapshot({
                stylelintInstallation: {
                    latest: expect.any(String)
                }
            });
        });
    });
});
