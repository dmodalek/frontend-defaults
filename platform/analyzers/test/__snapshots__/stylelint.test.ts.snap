// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers StylelintAnalyzer should analyze a project with installation correctly 1`] = `
Object {
  "stylelint": true,
  "stylelintIgnore": true,
  "stylelintInstallation": Object {
    "declared": "9.0.0",
    "installed": "9.10.1",
    "latest": Any<String>,
    "location": "devDependencies",
  },
}
`;

exports[`Analyzers StylelintAnalyzer should analyze a project without ignore & installation correctly 1`] = `
Object {
  "stylelint": true,
  "stylelintIgnore": false,
  "stylelintInstallation": Object {
    "declared": null,
    "installed": null,
    "latest": Any<String>,
    "location": null,
  },
}
`;
